go/ql/lib/semmle/go/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -322,6 +329,13 @@ class DataFlowCallable extends TDataFlowCallable {
   }
 }
 
+private Location getCallableLocation(Callable c) {
+  exists(string filepath, int startline, int startcolumn, int endline, int endcolumn |
+    c.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn) and
+    result.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+  )
+}
+
```

Comment:
It seems it would be equivalent to put this as a member predicate on `Callable`, and then it would be available to use from more places. Is there any downside to that?

Reply:
I did something like that first (in fact, my initial draft went a bit further and added even more `getLocation` member predicates), but there is a downside. Currently the Go type `Location` is the union of `DbLocation`, which is what you'd generally expect to be called `Location`, and some synthesized locations. The synthesized locations rely on `DataFlowNode.hasLocationInfo`, so the entire `Location` type is then dependent on a lot of data flow stuff. The current setup has been very carefully crafted to allow all of the regular ast classes to get locations without depending on anything data-flow-related. If we were to add `Location getLocation()` member predicates outside of data flow then that could break. To properly remedy this, then we would need to carefully go through the locations of all locatables, and see whether they could be given actual `@location`s and for those that need synthesized locations, then we'd need to check whether we could construct those in a way that doesn't depend on `DataFlow::Node`. So there's a lot of untangling to potentially do, and that was way beyond the scope of this PR.